### DISCUSSION TITLE
AWS Amplify Next.js app seems incompatible with NextAuth?

### DISCUSSION THREAD

**Discussion by ghost** [nextauthjs/next-auth#3103] [Help]
### Question 💬

After following the AWS Amplify Next.js tutorial [AWS Amplify Next.js](https://docs.amplify.aws/guides/hosting/nextjs/q/platform/js/#adding-amplify-hosting-1) I was able to successfully deploy an app. So then I decided to add Prisma and NextAuth to handle the authentication, I added Google as my OAuth provider. It works perfectly in development, I've tested it out.

However, when I pushed my code and it got redeployed to AWS I started getting this error

![issue](https://user-images.githubusercontent.com/93207292/140097285-a425c8e7-3cb2-4a4f-82b3-dd2f2e10232e.png)

So I am now wondering if there is a workaround for that.. In the AWS Amplify Next.js tutorial, at the bottom, they say the following:

`Amplify now supports API routes in Next.js apps. Any file inside the folder pages/api is mapped to /api/* and treated as an API endpoint instead of a page. You can use these APIs to interface with any backend service to fetch data.`

which makes me think that the issue comes from the fact that the NextAuth api handler's signature looks like this `[...nextauth].ts`

Has anyone tried this combination, I think if it works it would be the best setup I've been able to find around. Really hoping that there is a workaround tbh.

Regards!

### How to reproduce ☕️

1. follow the tutorial I've linked in the issue
2. add NextAuth to it

that's it, you get the aforementioned error

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by balazsorban44** [comment]
That's a standard Next.js feature, it not special to next-auth https://nextjs.org/docs/api-routes/dynamic-api-routes#catch-all-api-routes

the problem might be in your setup, although it's hard to tell without any reproduction.

**Reply by ghost** [reply]
So I guess then the problem comes somewhere from AWS / Amplify

PS: I will go ahead and create a reproducible example

**Reply by balazsorban44** [reply]
you could follow up there, yeah. one guess is to check that the `NEXTAUTH_URL` environment variable is set correctly

**Reply by ghost** [reply]
```NEXTAUTH_URL=https://main.d1munt2kh8plao.amplifyapp.com/``` this is it, I've copied the url from the deployed version in AWS

**Reply by ghost** [reply]
omg I just realized from your suggestion that I've completely forgotten to set these variables in the deployed version, a little rubber duck debugging moment right there lol, let me try it and I shall get back to you

**Reply by ghost** [reply]
@balazsorban44 [example repo](https://github.com/mallexplorer/next-amplify-next-auth-example) there you go - I've created a repo with my exact setup

1. created an app using ```npx create-next-app@latest --ts```
2. amplify init
3. followed [AWS Amplify Next.js Tutorial](https://docs.amplify.aws/guides/hosting/nextjs/q/platform/js/#adding-amplify-hosting-1) step by step
4. added Prisma (prisma init) and added NextAuth, configured it to work with Google
5. I've set all actual environment variables in the AWS Amplify app configuration Environment Variables and then for each lambda function I've set them in the lambda functions' environment variables configurations as well

**Comment by ghost** [comment]
seems that my concerns will turn up true - next auth actually treats some of these api routes as pages and AWS Amplify simply treats them as api routes and nothing else.. so I guess this won't work, I think I will have to use passport for that purpose

**Reply by ghost** [reply]
unless NextAuth provides a way to be used that is kind of headless

**Reply by balazsorban44** [reply]
>  next auth actually treats some of these api routes as pages

I'm not sure what you mean... `next-auth` **is** just API routes...?

It is more likely that Amplify might not support catch-all routes. The wording here makes me think this:

https://docs.amplify.aws/guides/hosting/nextjs/q/platform/js/#adding-amplify-hosting-1

> Any file inside the folder pages/api is mapped to /api/* and treated as an API endpoint instead of a page

So basically Amplify turns `/api/auth/[...nextauth]` into an API route literally.

You could test it out by creating a different route like `/api/test/[...slug].js` with this code:

```js
export default function test(req, res) {
	res.json({ slug: req.query.slug })
}
```

And try visiting:

1. `example.com/api/test/foo`
2. `example.com/api/test/foo/bar`

1 and 2 should return you the URL after `/test`. Is this the case? If not, then Amplify is the problem. I can give you a workaround if that is true, could you confirm first?

**Reply by ghost** [reply]
@balazsorban44  I followed ur instructions and indeed it returns ```{"slug":["foo"]}``` and ```{"slug":["foo", "bar"]}``` respectively, so what would the workaround be?

**Reply by balazsorban44** [reply]
Huh, then my workaround is void. :thinking: Do you have any logs in Amplify you could share? The error page you land on still seems to be a bit vague to pin the problem down.

There is nothing at all "fancy" in `next-auth` that wouldn't work on a compatible hosting provider if they support all the features Next.js requires.

**Reply by ghost** [reply]
I will try to get out some logs and supply them here a bit later today, however after looking at the source code of next-auth last nite I think that the problem seems to be that it does use some pages which end up as part of the ```/api``` route and given what Amplify says about treating everything in there as purely lambda function stuff it would suggest that whatever next-auth is trying to do that is OK with Next.js ends up being not OK within the AWS environment because of that.. so I guess there is no solution in terms of how next-auth approaches things in the first place, so I was thinking I will have to use passport and maybe go through your source code to see your approaches around jwt / csrf and that sort of thing and then implement my custom solution for it

**Reply by ghost** [reply]
maybe worth thinking of some sort of a headless solution that would simply be oriented around such an approach where it's purely backend related without any pages being injected in order to account for such environments in the future, I will be going through your source code a lot and if I manage to figure out a way to implement such a feature with your current architecture I will submit a PR, if not - I will just create a public repo with my custom solution once I've managed to get it working properly

**Reply by balazsorban44** [reply]
Could you elaborate on this, I am having trouble understanding:

> the problem seems to be that it does use some pages which end up as part of the /api route 

That doesn't make sense to me.

If you are thinking about the folder here https://github.com/nextauthjs/next-auth/tree/beta/src/core/pages, they are sent as regular HTML content-type responses, perfectly normal to do. In any case, you can override them https://next-auth.js.org/configuration/pages

**Reply by ghost** [reply]
yeah but there is still some react code involved, maybe that's what causing the issue, I am not sure either, just a suggestion I had.. I will investigate further tonite

**Reply by balazsorban44** [reply]
There is nothing else sent down to the client than plain HTML. JSX is transpiled away when the library is published as well: https://www.unpkg.com/browse/next-auth@4.0.0-beta.6/core/pages/error.js

As long as Amplify runs the required minimum Node version (12), all JS features in the package should be working as intended.

I'm still not convinced that this is a problem with `next-auth`

**Reply by ghost** [reply]
ok, I found out that my environment variables were not being passed to my lambdas, and after some struggling I finally managed to make them pass through.. However, even though they now work fine, I am still getting the same error, will continue my investigation tomorrow

**Reply by ghost** [reply]
so after all it turned out the combination with Prisma was causing the issue, I finally made it work! apparently something related to the build process with the Prisma dependencies was off, I've removed it for the time being and now it works! Thank you for your guidance @balazsorban44 

PS: good thing is I actually managed to solve the problem with the environment variables which I saw 2 issue threads on Github about and I've posted my solution there so at least something good came out of all that hassle

**Reply by Emily-Thomas** [reply]
@mallexplorer Hey, I am having an issue with env variables in my amplify setup with next auth. Can you link me to those github threads?

**Reply by bojidaryovchev** [reply]
@Emily-Thomas https://github.com/aws-amplify/amplify-console/issues/1987#issuecomment-961510561

**Reply by hwan27** [reply]
@ghost I'm having an issue similar to you using combination with Prisma. Could you explain more detail and how did you solve this?
I set my variables in amplify.yml(build setting) and still dosn't work.

**Reply by sorokinvj** [reply]
@ghost

> so after all it turned out the combination with Prisma was causing the issue, I finally made it work! apparently something related to the build process with the Prisma dependencies was off, I've removed it for the time being and now it works! Thank you for your guidance @balazsorban44
> 
> PS: good thing is I actually managed to solve the problem with the environment variables which I saw 2 issue threads on Github about and I've posted my solution there so at least something good came out of all that hassle

Congrats! did you need to specify any rewrites in Amplify Console Rewrites and redirects to make [...nextauth] work?

**Comment by BrandonEscamilla** [comment]
Hey everyone, I am having issues while deploying a next.js app using Next-auth in amplify, it seems the sign-in function is not working well using credentials provider. Is this related?

**Reply by danestves-reworth** [reply]
Having the same issue too

**Reply by balazsorban44** [reply]
This is most likely an Amplify compatibility issue. [Their docs site recites Next.js 10](https://docs.amplify.aws/guides/hosting/nextjs/q/platform/js), which has been out since 2020 October: https://nextjs.org/blog/next-10

**Reply by balazsorban44** [reply]
In any case, provide a reproduction with your code and possibly a deployed version so someone could help you out. (I don't have access to Amplify so wouldn't know how to deploy myself)

**Reply by BrandonEscamilla** [reply]
Hey @balazsorban44, I am sorry for posting here this issue with no more details, I am sure I can do better.  At the end the issue was my fault for not taking into account the environment variables and my poor debugging skills in AWS lambdas,  as far as I've seen this is an issue not related with Next.js or Next-auth, but Amplify. It is actually a very common issue since the documentation is somehow not very explicit with this topic, most persons try to set the environment variables only in the section to do that in their dashboard, but there is an issue with Amplify having problems to pass these environment variables to the lambdas functions, so u need to pass these variables also in the build process and create and .env file. I attached some images for visualization purpose. The fact is that is somehow difficult to debug the lambdas functions in AWS, so is difficult to see the errors in the server side end. Once more than 1 hour debugging I finally noticed that there was a very simple error, NEXTAUTH_URL was undefined. I think this should solve most issues developers are facing with Amplify and next-auth. The other one everyone is talking about is having issues using <Image/> tag since is not supported from Amplify yet. Hope this can be helpful for someone. 

![CleanShot 2022-03-01 at 10 48 27](https://user-images.githubusercontent.com/21966716/156146154-bd4585ea-8350-4396-b49e-33af045c4ce2.png)

![CleanShot 2022-03-01 at 10 47 05](https://user-images.githubusercontent.com/21966716/156146147-56751335-568c-49e7-8103-b8c0af965470.png)
.

**Reply by SawkaDev** [reply]
@BrandonEscamilla great post! however, shouldnt the npm run build come after the echo >> .env commands?

**Reply by foobarnes** [reply]
Hello all!

Just found this buried in Amplify's docs (classic): https://docs.aws.amazon.com/amplify/latest/userguide/ssr-environment-variables.html

It helped me make my NEXTAUTH_URL env variable available to my API routes.

**Reply by dShot7211** [reply]
foobarnes fix worked for me we have to configure server side variables in aws

**Comment by stvnbash** [comment]
I've gotten NextAuth to work on Amplify before.  I would recommend first testing your app on a hosted platform, such as Vercel.  Since Vercel makes NextJS, it is super easy to deploy your app to their platform for testing.  If it works there, redeploy to AWS.  Be sure to set up your environment variables properly.  You may also need to add an amplify.yml file to your project so Amplify passes your env variables into the application
amplify.yml will probably look something like this
```
version: 1
frontend:
  phases:
    preBuild:
      commands:
        - npm ci
    build:
      commands:
        - npm run build
        - echo "NEXTAUTH_SECRET=$NEXTAUTH_SECRET" >> .env
        - ... other env variables here ...
        - echo "NEXTAUTH_URL=$NEXTAUTH_URL" >> .env
  artifacts:
    baseDirectory: .next
    files:
      - '**/*'
  cache:
    paths:
      - node_modules/**/*
```

**Reply by shivpratik** [reply]
I did exactly what you said. I first deployed on Vercel and it works perfectly fine. When deployed on AWS Amplify the middleware cause issue. Only the protected routes give 500 error, all other pages work fine. If I remove the middleware all pages are working.

Below is my middleware code:

```javascript
import { withAuth } from "next-auth/middleware";
export default withAuth({
  secret: process.env.NEXTAUTH_SECRET,
});

export const config = {
  matcher: ["/dashboard"],
};
```  


When I check CloudWatch I see this error:

`Error: ENOENT: no such file or directory, open '/tmp/app/.next/server/edge-chunks/wasm_a7e702bc5ba9227243abda7977c8096f59478d3f.wasm'`

**Comment by gabrielfurini** [comment]
In my case, running the app in AWS Amplify, any of the methods to set environments on the build step worked without setting `AUTH_TRUST_HOST` as well. https://authjs.dev/getting-started/deployment#auth_trust_host

**Comment by z0ph** [comment]
Did the trick for me after 4 hours of debugging.

Add variables to the build phase of Amplify, in an .env file, to enforce the presence of the variable. (Was present in the Amplify Console, but was not sufficient)

**Comment by pawan-cloudcacheconsulting** [comment]
```
version: 1
frontend:
  phases:
    preBuild:
      commands:
        - npm ci
        - rm -rf .next
    build:
      commands:
        - npm run build
        - echo "NEXTAUTH_SECRET=$NEXTAUTH_SECRET" > .env.production
        - echo "NEXTAUTH_URL=$NEXTAUTH_URL" >> .env.production
        - echo "NODE_ENV=production" >> .env.production
        - echo "AUTH_TRUST_HOST=true" >> .env.production
        - echo "GOOGLE_CLIENT_ID=$GOOGLE_CLIENT_ID" >> .env.production
        - echo "GOOGLE_CLIENT_SECRET=$GOOGLE_CLIENT_SECRET" >> .env.production
        - echo "STRIPE_SECRET_KEY=$STRIPE_SECRET_KEY" >> .env.production
        
        # Build
        
        
        # Verify build
        - ls -la .next
  artifacts:
    baseDirectory: .next
    files:
      - '**/*'
  cache:
    paths:
      - node_modules/**/*
      - .next/cache/**/*
```


still its not working for me, i tried all best approaches, my middleware still not working....
